Clang tidy - Use Default for Constructors / Destructors

DavidFair opened the pull request:
**Description of work.**

Marks constructors and destructors with a default implementation as
default. This gives better diagnostic errors, and follows the principle of least astonishment.


**To test:**
- Check tests pass

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Manually fix namespaces in MantidVersion template

This brings the file in-line with out other modernised namespaces

DavidFair committed:
Clang-tidy =default c'tors and d'tors

Marks constructors and destructors with a default implementation as
default. This gives better diagnostic errors, and follows the principle
of least astonishment better.

martyngigg commented on Framework/API/src/Algorithm.cpp:
```diff
@@ -116,7 +116,7 @@ Algorithm::Algorithm()
 
 /// Virtual destructor
-Algorithm::~Algorithm() {}
+Algorithm::~Algorithm() = default;
```
Just out of interest. Is there any reason we need to define the constructor at all and not let the compiler generate it as it's `default`?

DavidFair replied:
In this case it's clang-tidy removing the stub implementation and replacing it with an explicit default

I can do a follow up PR to remove all instances of `T::~T() = default` (which don't override) in the codebase?

martyngigg replied:
Ah yes I see.

I think it might be worth it or someone else will likely come along with the same question as to why it's defaulted. This should have the bonus of inlining it?

martyngigg replied:
Do we know why this is marked override but it doesn't do anything? The base class `IPropertyManager.h` has a virtual destructor so this shouldn't need one?

martyngigg replied:
We've got quite a few with trivial overridden destructors that I can see in the changeset I think?

DavidFair replied:
Originally I tried removing them, but it resulted in several hundred linker errors. 
So ultimately it was easier for me to replace `~T() override {} = default;` and rely on future developers to notice that the destructor in both places are trivial and can be removed as/when they touch the code

DavidFair committed:
Remove stub destructors

Automatically removes various stub destructors and rely on the compiler
generated one.

DavidFair committed:
Mark ~T() override {} as default

Mark overriding destructors as default in headers instead.

DavidFair committed:
Mark virtual d'tors with empty impl default

Marks virtual ~T() {} as default, rather than using a stub
implementation

DavidFair committed:
Add missing virtual to destructors

These were found by searching for empty desctructors not marked virtual,
despite the fact the derive from a parent

DavidFair committed:
Remove extra destructors

These descructors arent:
- Virtual
- Overriding
- Non-trivial

So we can simply drop them alltogether

DavidFair committed:
Getting any remaning instances missing defaults

Re-run clang tidy to collect any remaining instances after our previous
changes

DavidFair committed:
Update Cppcheck Suppression Positions

Update to new positions since we've removed some destructors
